### ISSUE TITLE
Infinite loop

### ISSUE THREAD

**Issue by little-dude** [lefcha/imapfilter#105]
I just upgraded to 2.6.1 and imapfilter seems to enter an infinite loop:

```
S (3): * OK The Microsoft Exchange IMAP4 service is ready.
C (3): 1000 CAPABILITY
S (3): 1000 OK CAPABILITY completed.
C (3): 1001 STARTTLS
S (3): 1001 OK Begin TLS negotiation now.
C (3): 1002 CAPABILITY
S (3): 1002 OK CAPABILITY completed.
C (3): 1003 LOGIN "mylogin" *
S (3): 1003 OK LOGIN completed.
C (3): 1004 CAPABILITY
S (3): 1004 OK CAPABILITY completed.
C (3): 1005 NAMESPACE
S (3): 1005 OK NAMESPACE completed.
C (3): 1006 SELECT "INBOX"
S (3): 1006 OK [READ-WRITE] SELECT completed.
C (3): 1007 UID SEARCH ALL
S (3): 1007 OK SEARCH completed.
Cleaning mailbox
C (3): 1008 UID SEARCH UID 1407994:1407995,1408030,1408037:1408038,1408052,1408054,1408056,1408068:1408069,1408074,1408080,1408083,1408094,1408096,1408101:1408102,1408116:1408117,1408134,1408140,1408155,1408165:1408167,1408170,1446445:1446684 [...] SUBJECT "JIRA"

# then it repeats:
S (3): * OK The Microsoft Exchange IMAP4 service is ready.
C (3): 1011 CAPABILITY
S (3): 1011 OK CAPABILITY completed.
C (3): 1012 STARTTLS
S (3): 1012 OK Begin TLS negotiation now.
C (3): 1013 CAPABILITY
S (3): 1013 OK CAPABILITY completed.
C (3): 1014 LOGIN "mylogin" *
S (3): 1014 OK LOGIN completed.
C (3): 1015 CAPABILITY
S (3): 1015 OK CAPABILITY completed.
C (3): 1016 NAMESPACE
S (3): 1016 OK NAMESPACE completed.
C (3): 1017 SELECT "INBOX"
S (3): 1017 OK [READ-WRITE] SELECT completed.
C (3): 1018 UID SEARCH UID 1407994:1407995,1408030,1408037:1408038,1408052,1408054,1408056,1408068:1408069,1408074,1408080,1408083,1408094,1408096,1408101:1408102,1408116:1408117,1408134,1408140,1408155,1408165:1408167,1408170,1446445:1446684 [...] SUBJECT "JIRA"
```

The config:

```
options.timeout = 120
options.subscribe = true
options.expunge = true

account = IMAP {
    server = 'mail.something.com',
    username = 'mylogin',
    password = 'something'
}

inbox = account['INBOX']
sent = account['Sent Items']

sentFilter = function (msgs)
    msgs:move_messages(inbox)
end

cleanMailbox = function (msgs)
    print("Cleaning mailbox")
    res = msgs:contain_subject('JIRA') - msgs:contain_subject('[build]')
    res = res:is_older(2)
    res:delete_messages()
end

repeat
    repeat
        cleanMailbox(inbox:select_all())
        sentFilter(sent:select_all())
        os.execute("sleep 60")
    until not true
until not account.INBOX:enter_idle()
```

Sorry for the issue title being very vague. Feel free to edit to something more relevant.

Also, thank you for imapfilter :+1:

**Comment by lefcha** [comment]
From which version did you upgrade?

Also, can you run with `-d debug.out` and copy/paste the output written to the `debug.out` file?

**Comment by little-dude** [comment]
Previous version was 5.2.7.
Here is the debug:
https://gist.github.com/little-dude/a104bb1d3735efe97063

**Comment by little-dude** [comment]
I found very few resources about this `BAD Command Error. 10` and they suggest it's a server issue... The upgrade may have been a simple coincidence. I'll try to compile and install previous version to confirm.

https://github.com/mscdex/node-imap/issues/44
https://bugs.horde.org/ticket/10442

**Comment by little-dude** [comment]
I confirm the issue is not present in 2.5.7

**Comment by lefcha** [comment]
Sorry for the delay, I had the answer in mind, but I didn't get the time to write it here.

So since version 2.6 there's been some changes in meta-searching, and while IMAP servers should support receiving requests like the `1009 UID SEARCH [...]` that we see in your debug log, in practice it seems that some don't.  Unfortunately your server has problems coping with such long requests and that's why it replies with `1009 BAD Command Error. 10`, and then it aborts, and also unexpectedly closes the connection with ` * BYE Connection closed.`.  Imapfilter then tries to recover the connection and retry the last command and the server acts the same way again, and this continues, well... indefinitely.

So the solution to this problem is to avoid to use the new optimized meta-searching, and this can be done by putting a limit on how long a `SEARCH` request can be.  This is controlled by the `options.limit` configuration option described in the `imapfilter_config(5)` manual page, and a recommended value for the general case would be `50`.  With this value you basically get a behavior similar to how imapfilter worked in versions before 2.6.

So to sum it up, please set `options.limit = 50` in the beginning of your configuration file, and try again.

**Comment by little-dude** [comment]
:+1:  it works like a charm! Thank you very much :)
